### PR TITLE
[wip] [cleanup] Broken batchnorm test

### PR DESCRIPTION
**Pull request by sshleifer**
Currently: some batchnorm tests pass, but params don't update.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
Thanks for adding this!

**Comment by min-xu-ai** [comment]
@sshleifer I wonder what's have been discovered here? So far, my plan is to use BN in its own FSDP wrapper (full precision, no flattening) for now. Do you think BN can be more generally supported under FSDP?

**Comment by sshleifer** [comment]
>  Do you think BN can be more generally supported under FSDP?


- I think it's very likely to be possible and very unlikely to be easy. `running_mean`, which is a buffer, needs to be sharded.  More specifically, I expect that if the vertical model slice on a worker produces features of shape X, and the unsharded `bn.running_mean` is shaped X*num_workers, there will be an exception.
- Currently, FSDP does not shard buffers, and allows the caller to pass in a `buffer_dtype`, 
- I'd  google for other batch norm implementations  (not `nn.BatchNorm2d`) that support model parallelism.
- - I'd check if [deepspeed](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage3.py) can handle it, (and if so how).

Sorry I can't be more helpful!

**Comment by min-xu-ai** [comment]
This is super helpful already!
